pkg/errors/errors.go

```diff
@@ -0,0 +1,158 @@
+package errors
+
+import (
+	"errors"
+	"net/http"
+	"strings"
+
+	"github.com/marwan-at-work/vgop/semver"
+	"github.com/sirupsen/logrus"
+)
+
+// Kind enums
+const (
+	KindNotFound   = http.StatusNotFound
+	KindBadRequest = http.StatusBadRequest
+	KindUnexpected = http.StatusInternalServerError
+)
+
+// Error is an Athens system error.
+// It carries information and behavior
+// as to what caused this error so that
+// callers can implement logic around it.
+type Error struct {
+	// Kind categories Athens errors into a smaller
+	// subset of errors. This way we can generalize
+	// what an error really is: such as "not found",
+	// "bad request", etc. The official categories
+	// are HTTP status code but the ones we use are
+	// imported into this package.
+	Kind     int
+	Op       Op
+	Module   string
+	Version  string
+	Err      error
+	Severity logrus.Level
+}
+
+// Error returns the underlying error's
+// string message. The logger takes care
+// of filling out the stack levels and
+// extra information.
+func (e Error) Error() string {
+	return e.Err.Error()
+}
+
+// Op describes any independent function or
+// method in Athens. A series of operations
+// forms a more readable stack trace.
+type Op string
+
+// E is a helper function to construct an Error type
+// Operation always comes first, module path and version
+// come second, they are optional. Args must have at least
+// an error or a string to describe what exactly went wrong.
+// You can optionally pass a Logrus severity to indicate
+// the log level of an error based on the context it was constructed in.
+func E(op Op, args ...interface{}) error {
+	if len(args) == 0 {
+		panic("errors.E called with 0 args")
+	}
+	e := Error{Op: op}
+	for _, a := range args {
+		switch a := a.(type) {
+		case error:
+			e.Err = a
+		case string:
+			switch {
+			case isModule(a):
+				e.Module = a
+			case isVersion(a):
```

Comment:
Did I understand it right - if we would have to wrap every version into errors.V to get rid of this?

Reply:
I think it will be easy to forget to add errors.V/M so I would leave it as it is but maybe others have different opinion. I could live with both I guess.

Reply:
Unfortunately I had to remove `isModule` because there are third party modules with absolutely no `/` -- for example `go.opencensus.io` is a valid package on its own. So I had to replace it with `errors.M` -- And now that I've replaced it with `errors.M` we might as well do `errors.V`. You can look at how that looks in two of the errors. I think it's not too terrible. And we'll be able to remember because now `errors.E()` only takes a string for an error.